src/contracts/WebTestCase.php

```diff
@@ -0,0 +1,130 @@
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\Contracts\Test\Rest;
+
+use Ibexa\Contracts\Test\Core\IbexaKernelTestTrait;
+use Ibexa\Core\MVC\Symfony\Security\UserWrapped;
+use Symfony\Bundle\FrameworkBundle\KernelBrowser;
+use Symfony\Bundle\FrameworkBundle\Test\WebTestCase as SymfonyWebTestCase;
+use Symfony\Component\Security\Core\User\UserInterface;
+
+abstract class WebTestCase extends SymfonyWebTestCase
+{
+    use IbexaKernelTestTrait;
+
+    protected KernelBrowser $client;
+
+    protected function setUp(): void
+    {
+        $this->client = self::createClient();
+
+        $user = $this->getUser();
+        $this->client->loginUser($user);
+    }
+
+    private function getUser(): UserWrapped
+    {
+        $userService = self::getUserService();
+        $apiUser = $userService->loadUserByLogin('admin');
+        $symfonyUser = $this->createMock(UserInterface::class);
+        $symfonyUser->method('getRoles')->willReturn(['ROLE_USER']);
+
+        return new UserWrapped($symfonyUser, $apiUser);
+    }
+
+    /**
+     * Replaces dynamic values (like creation time) in $content with constant values.
+     */
+    private static function doReplace(string $content): string
+    {
+        $snapshotReplacer = static::getSnapshotReplacer();
+
+        return $snapshotReplacer->doReplace($content);
+    }
+
+    protected static function getSnapshotReplacer(): SnapshotReplacer
+    {
+        $replacer = new SnapshotReplacer();
+        $replacer->withJsonReplacements();
+        $replacer->withXmlReplacements();
+        $replacer->withJsonIgnoreIdChanges();
+        $replacer->withXmlIgnoreIdChanges();
+
+        return $replacer;
+    }
+
+    protected static function assertResponseMatchesXmlSnapshot(string $content, ?string $file = null): void
+    {
+        self::assertStringMatchesSnapshot($content, 'xml', $file);
+    }
+
+    protected static function assertResponseMatchesJsonSnapshot(string $content, ?string $file = null): void
+    {
+        self::assertStringMatchesSnapshot($content, 'json', $file);
+    }
+
+    /**
+     * @phpstan-param "xml"|"json"|null $type
+     */
+    protected static function assertStringMatchesSnapshot(
+        string $content,
+        ?string $type = null,
+        ?string $file = null
+    ): void {
+        $content = self::doReplace($content);
+
+        if ($file === null) {
+            $class = substr(static::class, strrpos(static::class, '\\') + 1);
+            $file = self::getDirectory() . '/_output/' . $class . '.' . ($type ?? 'log');
+        }
+
+        if (getenv('TEST_REST_GENERATE_SNAPSHOTS')) {
+            self::createDirectoryIfNotExists($file);
+
+            if (!file_exists($file)) {
+                file_put_contents($file, rtrim($content, "\n") . "\n");
+            }
+        }
+
+        if ($type === 'xml') {
```

Comment:
`xml` and `json` are good candidates for constants since they are heavily reused.